cmd/pinniped-proxy/src/logging.rs

```diff
@@ -0,0 +1,108 @@
+use std::cmp::PartialEq;
+use std::fmt;
+
+use hyper::{Body, Request, Response};
+
+#[derive(Clone, Debug, Default, PartialEq)]
```

Comment:
So here we're allowing four traits to be derived automatically:
* Clone enables an `object.clone()` method which, well, returns a clone of the struct. This becomes important in rust where the compiler helps you to be very explicit about ownership of variables as they get passed around. Cloning a var gives you a way to ensure the function you're passing the var into can own its own copy.
* Debug enables formatting the struct into a string using the `{:?}` format (e.g.: `info!("log data is: {:?}", log_data)`)
* Default enables the creation of a struct with sensible defaults (or you can implement it explicitly for different defaults)
* PartialEq enables equality testing of two instances of the struct (again, you can implement it explicitly for custom comparisons)

Reply:
> Debug enables formatting the struct into a string using the {:?} format (e.g.: info!("log data is: {:?}", log_data))

I don't know if I'm getting it: you are enabling the `Debug` trait, so you are letting the compiler implement the, let's say, toString() of this struct. However, you aren't using it in this file, aren't you?
I guess it is intended for other call-sites when you may need to log the value of a log_data object.

So, I suppose that when implementing a new struct, we always should add certain traits. Isn't there any way to ask the compiler to do so with every struct in the project? Don't know, just asking

Reply:
I believe it's used in this file. Clone in line 35, debug in any print (like in line 28) and default and PartialEq in the tests. What I don't know is how you map these `derive` to the functions used afterwards.

Reply:
> I believe it's used in this file. Clone in line 35, debug in any print (like in line 28) and default and PartialEq in the tests.

Yep, that's correct. The debug output formatter is the `{:?}` or `{:#?}`.

> What I don't know is how you map these `derive` to the functions used afterwards.

The compiler just adds implementations to the struct for them automatically. You can implement them manually (and have to for some, such as the Display trait, which can't be derived automatically since it depends or your own needs). More info at https://doc.rust-lang.org/book/ch10-02-traits.html

> So, I suppose that when implementing a new struct, we always should add certain traits. Isn't there any way to ask the compiler to do so with every struct in the project? Don't know, just asking

Others have asked too (especially about the debug trait, which seems obvious): https://github.com/rust-lang/prev.rust-lang.org/issues/1073